Add harbour for GZX. Teach hsct to get sources from git.

jxsvoboda opened the pull request:
We use XCW (not facade) for the actual build.

vhotspur commented:
Good to see another harbour opening :-). But I still have several comments...

I do not like the Git style. Commit 4d449088e1839e24a065f084c8320a85247ab9aa should have been split into two commits, the commit message does not follow Git conventions (length, dots).

Why not to download the GZX sources directly from GitHub as a tarball? The UI offers by default only ZIP, but rewriting the URL to end with `.tar.gz` works fine...

The Git origin etc. is not very well thought-through in my opinion. Why not to add it to `$shipsources` in a mini-syntax `git:REV:URL-PASSED-TO-GIT-CLONE`?

I do not understand the ping-pong you did with the return codes. Have you also modified the other shell script?

Is it possible to build GZX after Coastline work directory was initialized but HelenOS directory was rebuild for a different architecture (unable to check at the moment myself)?

Just my $0.02.

jxsvoboda commented:
Wait, there is a standard for writing git commit messages? And we're supposed to follow it? I had no idea! How is this possible? I don't recall any discussion over this... Sure I can fix it and re-structure the commits, but not sure if I can do it without opening a new pull request.

I didn't realize you can download the archive. But if you do, you always get the latest revision. Is it not preferable to get a specifc revision (that is know to work)? I thought about uploading a source release tarball and downloading that, but IMHO that is just overkill to do every time you want to bump the SW revision in HelenOS. What's wrong with git clone? As a bonus the content is verified with a hash.

I decided not to do a mini syntax for the following reasons:
 - it makes the implementation more complex
 - I couldn't think of a sensible way to support more than one git origin in the list
if you'd prefer the mini syntax, I can do it.

Return codes: Perhaps you could explain how those are supposed to work and how you are supposed to maintain those? I thought you wanted to have a separate error code for every return statement and you assign them sequentially as they appear in the code. That seems quite crazy compared to designing a few sensible error reasons, but perhaps there is some genius scheme... I didn't realize you consume the numbers somewhere else (because they don't look like an interface, just like a debugging feature). So what should be done here? You didn't quite explain...

Yes, you can build the harbour regardless of the state of HelenOS dir, as long as it exists - since commit "XCW should rely on export dir instead of touching the source tree. " everything is taken from the export dir (which the harbor specifies as the HSCT cache dir) and it only uses xcw tools from the source directory, which are not dependent on the state of the workspace.

jxsvoboda commented:
Why you cannot have more than one git origin: with git clone you need to specify the destination directory. If you don't, you get something unpredictable. We specify shipname as the destination dir and then use the same in the build dir and automatically chdir into it. If we had more origins, you'd somehow need to specify the names of the destination dirs in the syntax and you couldn't auto-chdir into.

vhotspur commented:
> Wait, there is a standard for writing git commit messages? And we're supposed to follow it? I had no idea! How is this possible? I don't recall any discussion over this... Sure I can fix it and re-structure the commits, but not sure if I can do it without opening a new pull request.

I only try to follow [our wiki](CodingTips) ([or here](http://www.helenos.org/wiki/GitCommitRules)) (and perhaps common sense to split independent changes into independent commits).

> I didn't realize you can download the archive. But if you do, you always get the latest revision.

No. When you [browse at specific commit](https://github.com/jxsvoboda/gzx/tree/65204bc836c7b53e313777bc64e1c594cf9289fd), you can download [tarball of that revision](https://github.com/jxsvoboda/gzx/archive/65204bc836c7b53e313777bc64e1c594cf9289fd.tar.gz).

> I decided not to do a mini syntax for the following reasons:
> - it makes the implementation more complex
> - I couldn't think of a sensible way to support more than one git origin in the list
if you'd prefer the mini syntax, I can do it.

Okay.

> Return codes: Perhaps you could explain how those are supposed to work and how you are supposed to maintain those? I thought you wanted to have a separate error code for every return statement and you assign them sequentially as they appear in the code. That seems quite crazy compared to designing a few sensible error reasons, but perhaps there is some genius scheme... I didn't realize you consume the numbers somewhere else (because they don't look like an interface, just like a debugging feature). So what should be done here? You didn't quite explain...

I think they are somehow used in the `build_all.sh` but I never understood why that exists. Perhaps @le-jzr can explain as he added it AFAIK.

> Yes, you can build the harbour regardless of the state of HelenOS dir, as long as it exists - since commit "XCW should rely on export dir instead of touching the source tree. " everything is taken from the export dir (which the harbor specifies as the HSCT cache dir) and it only uses xcw tools from the source directory, which are not dependent on the state of the workspace.

Thanks, as I wrote, I was unable to check that.

le-jzr commented:
> Return codes: Perhaps you could explain how those are supposed to work and how you are supposed to maintain those? I thought you wanted to have a separate error code for every return statement and you assign them sequentially as they appear in the code. That seems quite crazy compared to designing a few sensible error reasons, but perhaps there is some genius scheme... I didn't realize you consume the numbers somewhere else (because they don't look like an interface, just like a debugging feature). So what should be done here? You didn't quite explain...

There's no great scheme. As you say, I did the unique numbering to help me debug the script. The sequential nature is inconsequential. :)
`build_all.sh` uses the return code, so you should update it as well. In hindsight, that should have been documented, sorry.

Commit:
Teach HSCT to download sources from git

Commit:
Add harbour for GZX

We set up XCW and let the upstream makefile use it as with manual
XCW build. Facade is not used.

Commit:
Add harbour for Sycek

vhotspur commented on hsct.sh:
```diff
@@ -193,7 +222,9 @@ hsct_fetch() {
+	rm -rf "$HSCT_BUILD_DIR/$shipname"
+	rm -f "$HSCT_BUILD_DIR/${shipname}.built"
+	rm -f "$HSCT_BUILD_DIR/${shipname}.packaged"
 }
```
May I ask why this is hidden in this commit? The purpose of the `clean` command is to save space but definitely not force recompilation (e.g. when compiling dependent packages).  BTW, I use the following patch to CI on the build machine to save space (yes, it matters) and now I see that all dependencies are always rebuild... Any particular reason for this? And yes, I admit this is not documented anywhere.

```diff
--- a/hbuild/builders/coastline.py
+++ b/hbuild/builders/coastline.py
@@ -202,6 +202,9 @@ class CoastlineBuildTask(Task):
         res = self.ctl.run_command([ root + '/hsct.sh', 'archive', self.harbour ], cwd=my_dir)
         if res['failed']:
             return False
+
+        # Clean build directory
+        self.ctl.run_command([ root + '/hsct.sh', 'clean', self.harbour ], cwd=my_dir)
         
         # Add downloadable archive
         profile_flat = self.profile.replace("/", "-")

```

jxsvoboda replied:
The replacement of rm -rf "$HSCT_BUILD_DIR/$shipname/"* with rm -rf "$HSCT_BUILD_DIR/$shipname" was needed by my original version of git support because the root of the git workspace was in $shipname and rm -rf ../* did not remove .git. Now there is an extra subdir so it is no longer necessary (but perhaps does not break anything).

Ad the  removal of .built and .packaged: If you perform "hsct.sh package pkg" and "hsct clean pkg" and "hsct package pkg", it will fail because it won't rebuild the package (since .built is present) and tries to package right away - but the built files are missing.

I thought it made sense to rebuild after a clean since there wasn't a way to force a rebuild nor a way to detect if a rebuild was needed.

I guess if we removed just $shipname.built, then it would fix the problem with package-clean-package and a build-clean-build would re-build the package (so that would be a way of forcing the package to be re-built; or, we could add another command like clobber to remove .packaged as well).

jxsvoboda replied:
And it should prevent the dependencies to be rebuilt during a archive-clean-archive cycle, for example.

vhotspur replied:
> The replacement of rm -rf "$HSCT_BUILD_DIR/$shipname/"* with rm -rf "$HSCT_BUILD_DIR/$shipname" was needed by my original version of git support because the root of the git workspace was in $shipname and rm -rf ../* did not remove .git. Now there is an extra subdir so it is no longer necessary (but perhaps does not break anything).

That is fine, not sure why I put the weird asterisk syntax there.

> Ad the removal of .built and .packaged: If you perform "hsct.sh package pkg" and "hsct clean pkg" and "hsct package pkg", it will fail because it won't rebuild the package (since .built is present) and tries to package right away - but the built files are missing.

Just tested it and that works fine. When the `.packaged` file is present, it spits out `>>> No need to package pkg`.

> I thought it made sense to rebuild after a clean since there wasn't a way to force a rebuild nor a way to detect if a rebuild was needed.

Aha, that makes sense too :-). I always removed whole `build/` for a clean rebuild. Now, I use CI for that as it tests for all architectures with single command and use `clean` just to save disk space after successful build.

To make both sides happy, I will add `distclean` (or `scrub`?) for complete clean-up of a package and keep `clean` to remove only the build directory (and refuse `clean` when `.built` but not `.packaged`).

jxsvoboda replied:
Oops, the sequence to reproduce the problem is (starting with a clean state) hsct.sh build pkg, hsct.sh clean pkg, hsct.sh package pkg.

vhotspur replied:
Should be fixed by eb0c5f21c3e5cf5077fe6b49f569e52af8dce266.

vhotspur commented:
Thanks for rewriting all of this! Especially the extra work for the GIT thing :-)
